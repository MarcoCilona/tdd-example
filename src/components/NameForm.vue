<template>
  <div>
    <input data-test-id="inputTest" v-model='name' />
    <button data-test-id="buttonTest" @click='handleClick'>Click me!</button>
  </div>
</template>

<script lang="ts">
import {
  Component, Vue,
} from 'vue-property-decorator';

@Component
export default class NameForm extends Vue {
  name = '';

  handleClick() {
    this.$emit('submit', this.name);
    this.name = '';
  }
}
</script>
