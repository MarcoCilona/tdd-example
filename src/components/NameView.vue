<template>
  <div>
    <p data-test-id="paragraphTest">Hello {{ name }}</p>
    <button data-test-id='greetingsButton'
      @click='handleMoreGreetingsClick'>More greeting please!</button>
  </div>
</template>

<script lang="ts">
import {
  Component, Prop, Vue,
} from 'vue-property-decorator';

@Component
export default class NameView extends Vue {
  @Prop() name!: string;

  handleMoreGreetingsClick() {
    this.$router.push({
      path: `/more-greetings/${this.name}`,
    });
  }
}
</script>
