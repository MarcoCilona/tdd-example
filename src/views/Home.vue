<template>
  <div>
    <NameForm @submit='handleSubmit'/>
    <NameView :name='name' />
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

// Components
import NameForm from '@/components/NameForm.vue';
import NameView from '@/components/NameView.vue';

@Component(
  {
    components: {
      NameForm,
      NameView,
    },
  },
)
export default class Home extends Vue {
  name = '';

  handleSubmit($event: string) {
    this.name = $event;
  }
}
</script>

<style>
</style>
