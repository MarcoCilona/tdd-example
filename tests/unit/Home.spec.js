import { shallowMount } from '@vue/test-utils';
import Home from '@/views/Home.vue';

//  Components
import NameForm from '@/components/NameForm.vue';
import NameView from '@/components/NameView.vue';

/**
 * Test needed:
 * 1. Componente Home exists and its the app container;
 * 2. Home component contains NameForm and NameView components
 * 3. Home saves the name retrieved from NameForm and saves it inside a variable
 * 4. Home passes the name saved to NameView component
 */
describe('Home tests', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Home, {
      data: () => ({
        name: '',
      }),
    });
  });

  it('Home component exists and is rendered right', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Home component contains NameForm component', () => {
    expect(wrapper.findComponent(NameForm).exists()).toBe(true);
  });

  it('Home component contains NameView component', () => {
    expect(wrapper.findComponent(NameView).exists()).toBe(true);
  });

  it('Home save the name retrieved from NameForm', async () => {
    const nameFormWrapper = wrapper.findComponent(NameForm);
    const nameToBeSaved = 'Digitiamo';
    nameFormWrapper.vm.$emit('submit', nameToBeSaved);

    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$data.name).toEqual(nameToBeSaved);
  });

  it('Home is passing the name to NameView component', async () => {
    const nameViewWrapper = wrapper.findComponent(NameView);
    const nameToBeSaved = 'Digitiamo';
    wrapper.setData({
      name: nameToBeSaved,
    });

    await wrapper.vm.$nextTick();
    expect(nameViewWrapper.props().name).toEqual(wrapper.vm.$data.name);
  });
});
