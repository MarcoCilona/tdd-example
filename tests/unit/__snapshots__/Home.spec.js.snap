// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home tests Home component exists and is rendered right 1`] = `
<div>
  <nameform-stub></nameform-stub>
  <nameview-stub name=""></nameview-stub>
</div>
`;
